### ISSUE TITLE
AbuseIPDB integration

### ISSUE THREAD

**Issue by 737simpilot** [CIDRAM/CIDRAM#89]
Is it possible to use the API to integrate with CIDRAM?  https://www.abuseipdb.com/api.html

I see that they have a CIDR API that may be what we are after. Or perhaps it would just be the Check IP API.

**Comment by 737simpilot** [comment]
I wonder if the module could be crafted in such a way to bypass an individual IP against the AbuseIPDB database should one complain that they are being blocked?

**Comment by ghost** [comment]
I would be wary to implement a solution 

> that requires that users register an account there first, if they want to submit a removal request, and it must be done from the actual page which displays the reports

especially since they also has shown less willingness to reply on inquiries.

The idea is basically good, but I can foresee issues where we or the CIDRAM end user get the blame for not being able to remove an IP in case of a legit removal request.

**Comment by Maikuolan** [comment]
Will delay a final decision for the moment and come back to to this properly a little later (probably within the next few days). My current priorities are finished off work on everything which has already been started, before starting work on any new modules, features and such, so that I can hopefully tag/release the next version (v1.8.0) at some point during my spare time reasonably soon, and be able to confidently say that there's nothing incomplete or half-finished in that version, but I'm also sorting out a job transition and other things outside the projects at the moment, which has been keeping me a little busy the past day or so, and probably will continue for at least another day or so (so there's not much free "head space" at this very moment, if that makes sense, but there should be again soon).

**Comment by 737simpilot** [comment]
Okay. Quick question: You don't plan to use PHP 7.2 anytime soon, do you? Because I have a phpBB extension that uses Amazon S3 and it doesn't work with PHP 7.2 at the moment until the author updates it which I'm hoping is withen the next few months.

**Comment by Maikuolan** [comment]
In my dev environment, where I tend to mostly work on CIDRAM, I actually already use the latest versions of all PHP branches from PHP 5.4 onward, which includes the latest version of PHP 7.2. Though, whether I plan to "use" it probably isn't really what you're wanting to know, but rather, whether I plan to set it as the minimum version at any time soon.

To answer that: No, I don't plan on upping the minimum required PHP version for CIDRAM any time *soon*. I do plan on doing so *eventually*, at some point in the distant future, but not in the near future. Upping the minimum required PHP version would be considered a backwards-incompatible change, which would necessitate also upping CIDRAM's major version number, thus any plans to up the minimum required PHP version wouldn't be happening until at least CIDRAM v2.0.0, and I don't anticipate any need to release v2.0.0 for at least another year or so. I do, however, already have some vague plans for an eventual v2.0.0 release, and some basic ideas on what I'd like to see in an eventual v2.0.0, so I anticipate that CIDRAM v2.0.0 should be released *eventually*, but not soon. In other words: There's plenty of time, and no need to worry yet.

On a side-note: PHP 7.4 (not yet released, but will probably be released late next year or the year after, I suspect) already has some documentation available about what the PHP team plans to deprecate, what won't be backwards-compatible and so on. Based on that documentation, no currently supported tagged versions of CIDRAM will be entirely compatible with PHP 7.4 (some small potential type-casting problems exist in CIDRAM which would affect PHP 7.4 compatibility), but the problem has already since been resolved in the codebase, so, the next released CIDRAM version (v1.8.0, to hopefully be released within the next few days or earlier) will be fully compatible.

**Comment by 737simpilot** [comment]
Okay, thanks.

**Comment by 737simpilot** [comment]
Any updates?

**Comment by 737simpilot** [comment]
Test

**Comment by 737simpilot** [comment]
test

**Comment by 737simpilot** [comment]
test

Sorry about the alert here. I had an issue with an add-on in my browser and using the edit button. Found the issue and resolved it.

**Comment by 737simpilot** [comment]
Thank you!

**Comment by 737simpilot** [comment]
Any development news? I really would love this added ability. LOL

**Comment by Maikuolan** [comment]
Not yet. Been busy with other tasks. Probably soon though.

**Comment by 737simpilot** [comment]
Okay, thanks.

**Comment by 737simpilot** [comment]
Do you have plans for a fail safe where if the AbuseIPDB API fails then the request is sent through anyway? 

Also, an IP bypass might be considered in case one were to get blocked illegitimately, but I doubt that will happen. Probably in rare cases, but one could also go to the AbuseIPDB website and request their IP removed as well.

**Comment by 737simpilot** [comment]
Any progress?

**Comment by DanielRuf** [comment]
As you can see there is no further progress currently and I see no dedicated feature branch so far.

**Comment by eurobank** [comment]
AbuseIPDB has a % of CONFIDENCE for each IP. I have no idea if this also available in their api, but if yes then it could be used. If something has 75% or greater then ...

**Comment by Maikuolan** [comment]
Good point.. I'll check that out again soon. A confidence score could (at least partially) resolve the problem here, I think.

**Comment by 737simpilot** [comment]
I don't know about confidence scores and what not, but what I do know is that every IP that slipped by CloudFlare and CIDRAM was listed at AbuseIPDB which pisses me off quite frankly. Yet they say we are running out of IPv4 address space??? HA! I just massaged a bunch of IPs from CIDRAM to add to my Peerblock lists and there are TONS! Caleb, I have no idea how you keep up. LOL

The one thing though that I don't much care for is that AbuseIPDB requires a paid subscription for a higher tier or what ever it was now. So that also could play a factor in "confidence scores."

**Comment by Maikuolan** [comment]
> The one thing though that I don't much care for is that AbuseIPDB requires a paid subscription for a higher tier or what ever it was now. So that also could play a factor in "confidence scores."

Good point. I would hope that it doesn't play a factor in their confidence scores, but I haven't looked into that yet, so.. something I'll definitely look into too.

**Comment by Maikuolan** [comment]
Definitely, a future AbuseIPDB module should (and will) utilise confidence scores; That'll help quite a bit with false positives, bad reports, abuse, etc (especially seeing as AbuseIPDB allows unregistered, anonymous reporting).

As a quick update: The reason I haven't gotten further with this yet (as it is currently), is because (1) the latest version of the AbuseIPDB API (v2) requires an API key to perform lookups (and using their v1 API isn't a viable long-term solution, because they've already marked it as deprecated and state in its documentation that they'll remove support for it entirely early 2020), and (2) in order for a future CIDRAM AbuseIPDB module to work correctly, a mechanism will need to be provided to allow users to supply this API key to CIDRAM somehow, but I haven't quite figured out what that should look like yet.

While simply adding a new configuration option for the API key may seem like the easiest and most obvious solution, doing that would somewhat break modularisation, in the sense that we're looking at creating a module, rather than an wholly integrated component, and I don't think that modules should rely upon the existence of dedicated, module-specific configuration in order to work correctly, nor should anything else specific to a particular module be originating from the core itself (i.e., other than within the `modules.dat` file, which is needed by the front-end updates page to be able to know which modules are available and how to install/update them, there shouldn't be any traces of any particular module remaining within the core CIDRAM installation after a user uninstalls a module, or when a user chooses to not install a particular module in the first place). Instead, adhering to proper modularisation, modules should be able to work correctly without needing any special treatment from the core, and should otherwise include anything they need to work correctly themselves, but the core should still nonetheless support modules by providing any non-specific mechanisms required by modules to work correctly in general (e.g., like how CIDRAM provides mechanisms to use `$Trigger()` and `$Bypass()` in modules).

However, needing an API key to perform lookups is not a requirement unique to AbuseIPDB. It's very common, in fact, that many useful and popular services need an API key to perform lookups or to otherwise leverage whatever service is being provided.

I was thinking about maybe writing a "key ring" feature for CIDRAM; A way that CIDRAM could store and fetch configurable data (such as API keys), separate from the configuration itself. But I haven't quite yet figured out how to best approach that, to ensure that it would remain secure (i.e., not likely to result in API keys being compromised due to exposure), and not super messy (i.e., we've already got configuration, the cache, auxiliary rules, signature files, and a bunch of other places where data can be stored for various reasons; adding yet another such place, or more, could start looking messy, or become confusing to some end-users over time, and we should aim to avoid such confusion as best as possible, I think).

**Comment by 737simpilot** [comment]
Thanks for the update.

**Comment by Maikuolan** [comment]
Figured out a good way to do this. Will begin working on this shortly.

**Comment by 737simpilot** [comment]
Cool!

**Comment by 737simpilot** [comment]
Reading the FAQs at AbuseIPDB, they have this:

> APIv1: The rate limiter is a 24-hour sliding window. In simple terms, if you hit your limit of 1,000 requests, and your first request was made 23 hours ago, then in one hour you gain that one request back. So on and so forth for each nth request. All requests that exceeded the limit will be served will a 429 Too Many Requests response. 

So I guess from that if I go over 1,000 requests in 24 hours using their free plane, my site will issue a 429? Or is that withen the API its self or something?

**Comment by Maikuolan** [comment]
> So I guess from that if I go over 1,000 requests in 24 hours using their free plane, my site will issue a 429? Or is that withen the API its self or something?

They're referencing API behaviour (in relation to rate limiting API usage; but it has no bearing on rate limits for end-users visiting the websites themselves that use the API). It means that if more than *n* API requests are made within a 24-hour period, the API will issue a 429 status code in response to any subsequent API requests made, until the sliding window shifts enough to bring the total number of requests made within the given period back to below *n* again (the actual *n* quota can vary, depending on whether using their free plan or a paid plan, but the free plan, which I assume that ~99% of CIDRAM users will end up using, allows for 1,000 requests).

We'll also be using their v2 API, rather than their v1 API. Official "sunset" date for their v1 API is towards the beginning of 2020, meaning that it won't be around for much longer anyway (i.e., *<strike>dead man</strike> dead API walking..*). Their v2 API is basically identical anyway though in regards to quotas.

**Comment by Maikuolan** [comment]
(Seems to be basically working properly now at my end anyway. Just wanting to do a bit more testing before I commit it or push it live. Should be up soon though, possibly within the next few hours or thereabouts).

**Comment by Maikuolan** [comment]
Done. :-)

The AbuseIPDB is now fully functional and available via the front-end updates page. You'll need to update to the latest available version of CIDRAM first (`=> 1.13.1-DEV+19206001`), or else the updater will refuse to install the AbuseIPDB module, due to the minimum version requirement (some changes implemented with the latest CIDRAM update that are necessary for the AbuseIPDB module to be able to work properly). Once you've updated though, you should be able to install it and activate it via the front-end updates page.

After activating it, some new configuration directives will appear on the configuration page (these new configuration directives won't be visible while the module is inactive). One of these configuration directives is for you to specify an AbuseIPDB API key. You'll need to fill that in, otherwise the module won't do anything and will basically be useless.

You can also optionally report bad things that CIDRAM detects back to AbuseIPDB, too. This is disabled by default though, so you'll need to enable it via the configuration page if you want that to happen.

**Comment by Maikuolan** [comment]
(Marking as implemented. Keeping the issue open for a while longer though, because I'll need to update the documentation, to mention about the new module, due to GDPR/PII TL;DR legal stuff that nobody's ever going to read anyway but that still needs to be written down nonetheless. Everything else here has been done now though).

**Comment by Maikuolan** [comment]
Finished updating docs. Closing.
